panel/widgets/chat.py

```diff
@@ -0,0 +1,1085 @@
+        returned entry is passed back into the method, e.g.
+        `entry = chat.stream(token, entry=entry)`.
+
+        Unlike send, this will not automatically execute the callback
+        upon completion; to do so manually, invoke the `respond` method.
+
+        This method is primarily for outputs that are not generators--
+        notably LangChain. For most cases, use the send method instead.
+
+        Parameters
+        ----------
+        value : str
+            The new token value to stream.
+        user : Optional[str]
+            The user to stream as; overrides the entry's user if provided.
+        avatar : Optional[Union[str, BinaryIO]]
+            The avatar to use; overrides the entry's avatar if provided.
+        entry : Optional[ChatEntry]
+            The entry to update.
+
+        Returns
+        -------
+        The entry that was updated.
+        """
+        replace = entry is None
+        entry = self._build_entry(entry or value, user=user, avatar=avatar)
+        if replace:
+            self._replace_placeholder(entry)
+        else:
+            self._stream(value, entry)
+        return entry
+
+    def respond(self):
+        """
+        Executes the callback with the latest entry in the chat log.
+        """
+        self._callback_trigger.param.trigger("clicks")
+
+    def undo(self, count: Optional[int] = 1) -> List[Any]:
+        """
+        Remove the last `count` of entries from
+        the chat log and returns them.
+
+        Parameters
+        ----------
+        count : Optional[int]
+            The number of entries to remove, starting from the last entry.
+
+        Returns
+        -------
+        The entries that were removed.
+        """
+        if count == 0:
+            return []
+        entries = self._chat_log.objects
+        undone_entries = entries[-count:]
+        self._chat_log.objects = entries[:-count]
+        return undone_entries
+
+    def clear(self) -> List[Any]:
+        """
+        Clears the chat log and returns
+        the entries that were cleared.
+
+        Returns
+        -------
+        The entries that were cleared.
+        """
+        cleared_entries = self._chat_log.objects
+        self._chat_log.clear()
+        return cleared_entries
+
+    @property
+    def entries(self) -> List[ChatEntry]:
+        """
+        Returns the entries in the chat log.
+        """
+        return self._chat_log.objects
+
+    @entries.setter
+    def entries(self, entries: List[ChatEntry]) -> None:
+        """
+        Sets the entries in the chat log.
+        """
+        self._chat_log.objects = entries
+
+
+class ChatInterface(ChatFeed):
+    """
+    High level widget that contains the chat log and the chat input.
+    """
+
+    widgets = param.List(doc="""
+        Widgets to use for the input. If not provided, defaults to
+        `[TextInput]`.""")
+
+    user = param.String(default="You", doc="Name of the ChatInterface user.")
+
```

Comment:
I'm lacking an `assistant` parameter that holds the name of the `assistant`. Currently the response generated by the callback becomes 'user' which does not make sense.

![image](https://github.com/holoviz/panel/assets/42288570/b483856f-e49a-4ad1-8464-18aaac200acb)

```python
import openai
import panel as pn
pn.extension(sizing_mode="stretch_width")


def callback(contents: str, user: str, instance: pn.widgets.ChatInterface):
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": contents}],
    )
    yield response.choices[0]["message"]["content"]

ci = pn.widgets.ChatInterface(callback=callback)
ci.servable()
```

Reply:
You can customize it as you'd like.
```python
import openai
import panel as pn
pn.extension(sizing_mode="stretch_width")


def callback(contents: str, user: str, instance: pn.widgets.ChatInterface):
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": contents}],
    )
    yield {"user": "assistant", "value": "response.choices[0]["message"]["content"]}

ci = pn.widgets.ChatInterface(callback=callback)
ci.servable()
```

Reply:
That is great. This will enable more than a 'user' and and 'assistant' chatting. But I would still recommend the default user for the callback to be called 'assistant' and not 'user'. Its confusing.

Reply:
Added `callback_user` and `callback_avatar`